src/java.naming/share/classes/com/sun/jndi/ldap/sasl/LdapSasl.java

```diff
@@ -123,15 +125,23 @@ public static LdapResult saslBind(LdapClient clnt, Connection conn,
         try {
             // Prepare TLS Channel Binding data
             if (conn.isTlsConnection()) {
-                TlsChannelBindingType cbType =
-                        TlsChannelBinding.parseType(
+                TlsChannelBindingType cbType;
+                try {
+                    cbType = TlsChannelBinding.parseType(
                                 (String)env.get(TlsChannelBinding.CHANNEL_BINDING_TYPE));
+                } catch (ChannelBindingException e) {
+                    throw new NamingException(e.getMessage());
+                }
                 if (cbType == TlsChannelBindingType.TLS_SERVER_END_POINT) {
                     // set tls-server-end-point channel binding
                     X509Certificate cert = conn.getTlsServerCertificate();
                     if (cert != null) {
-                        TlsChannelBinding tlsCB =
-                                TlsChannelBinding.create(cert);
+                        TlsChannelBinding tlsCB;
+                        try {
+                            tlsCB = TlsChannelBinding.create(cert);
+                        } catch (ChannelBindingException e) {
+                            throw new NamingException(e.getMessage());
+                        }
```

Comment:
Should we call initCause here and above? I see that we do call initCause in NegotiatorImpl.java.
On the one hand it's better for diagnostic. On the other hand it exposes a module-internal exception class which is not great. Or maybe we should set the cause of the CBE as the cause of NamingException.

Reply:
As long as the spec has not dictated what the cause should be, I don't care if the exception type is internal or not.